spark-3/src/main/scala/org/neo4j/spark/DataSource.scala

```diff
@@ -38,23 +42,42 @@ class DataSource extends TableProvider
 
   private val jobId: String = UUID.randomUUID().toString
 
-  private var schema: StructType = null
+  private var schema: StructType = _
 
-  private var neo4jOptions: Neo4jOptions = null
+  private var neo4jOptions: Neo4jOptions = _
+
+  private var neo4j: Neo4j = _
 
   override def supportsExternalMetadata(): Boolean = true
 
   override def inferSchema(caseInsensitiveStringMap: CaseInsensitiveStringMap): StructType = {
     if (schema == null) {
       val neo4jOpts = getNeo4jOptions(caseInsensitiveStringMap)
       Validations.validate(ValidateConnection(neo4jOpts, jobId))
+      val neo4j = getNeo4jInfo(neo4jOpts.connection)
       schema =
-        Neo4jUtil.callSchemaService(neo4jOpts, jobId, Array.empty[Filter], { schemaService => schemaService.struct() })
+        Neo4jUtil.callSchemaService(
+          neo4j,
+          neo4jOpts,
+          jobId,
+          Array.empty[Filter],
+          { schemaService =>
+            schemaService.struct()
+          }
+        )
     }
 
     schema
   }
 
+  private def getNeo4jInfo(options: Neo4jDriverOptions): Neo4j = {
+    if (neo4j == null) {
+      val driver = new DriverCache(options).getOrCreate()
+      neo4j = Neo4jDetectorKt.detectedWith(Neo4j.Companion, driver)
```

Comment:
Have you tried `@JvmStatic` and `@JvmName` annotations on Kotlin side, so that we get rid of `Kt` or `Companion` uses?

Reply:
If I do that, it will slightly degrade the experience in Kotlin, since `detectedWith` is defined as an extension function and called like this: `Neo4j.detectedWith(driver)`.